### PR TITLE
FIPS 140-3 changes

### PR DESCRIPTION
**Pull request by paulidale**
This represents the majority of the changes that are currently known to be required for a FIPS 140-3 validation.

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by mattcaswell** [comment]
CI failures look relevant

File: providers/fips/self_test_kats.c
Comment (beldmit) [inline providers/fips/self_test_kats.c:852]: We considered the approach with predictable DRBG but considered it quite risky because an implementation error will cause using the predictable DRBG outside of the KAT context.

File: providers/fips/self_test_kats.c
Comment (hlandau) [inline providers/fips/self_test_kats.c:699]: This seems to bypass the test (returning 1) if `set_kat_drbg` fails?

**Review by hlandau — COMMENTED** [review]
Looks a lot less painful than I was expecting!

File: providers/fips/self_test_kats.c
Comment (paulidale) [reply providers/fips/self_test_kats.c:699]: Yeah, it should fail...

File: providers/fips/self_test_kats.c
Comment (paulidale) [reply providers/fips/self_test_kats.c:852]: Yes, that is a concern.  I don't see a nice way around this risk.

File: providers/fips/self_test_kats.c
Comment (paulidale) [reply providers/fips/self_test_kats.c:699]: Added a pile of error handling and recovery.

File: test/recipes/03-test_fipsinstall.t
Comment (paulidale) [inline test/recipes/03-test_fipsinstall.t:289]: I'm unsure as to supporting this here.  This and the modified block further down is verifying that the self check on install works.  This is removed in 140-3, should these tests also be removed?

I'm leaning towards _yes_ to avoid clutter here.  This would also let us remove some of the config support for the setting.

File: crypto/dsa/dsa_key.c
Comment (slontis) [inline crypto/dsa/dsa_key.c:83]: == NULL

File: crypto/dsa/dsa_key.c
Comment (slontis) [inline crypto/dsa/dsa_key.c:163]: indent 1 more?

File: crypto/ec/ec_key.c
Comment (slontis) [inline crypto/ec/ec_key.c:273]: Why does the dirtycount increment?

File: crypto/ec/ec_key.c
Comment (slontis) [inline crypto/ec/ec_key.c:260]: == NULL

File: crypto/dsa/dsa_key.c
Comment (slontis) [inline crypto/dsa/dsa_key.c:96]: sizeof(bytes) is not used anywhere.... Same comment applies to next file..

File: crypto/ec/ec_key.c
Comment (slontis) [inline crypto/ec/ec_key.c:546]: I am not seeing what this comment is referring to in the IG. please update if it is not correct,

File: doc/man3/RAND_get0_primary.pod
Comment (slontis) [inline doc/man3/RAND_get0_primary.pod:71]: on the

Comment (slontis) [reply doc/man3/RAND_get0_primary.pod:71]: or 'apply only on'

File: providers/fips/self_test.c
Comment (slontis) [inline providers/fips/self_test.c:179]: How were these values generated?

File: crypto/dsa/dsa_key.c
Comment (paulidale) [reply crypto/dsa/dsa_key.c:96]: Hmmmm, not a bug just a problem waiting to happen...

Still easy to return an error if the buffer isn't large enough.

File: providers/fips/self_test.c
Comment (slontis) [inline providers/fips/self_test.c:183]: Maybe plonk this in the .inc file?

File: providers/fips/self_test.c
Comment (paulidale) [reply providers/fips/self_test.c:179]: No idea.  KeyPair did them.

File: providers/fips/self_test_kats.c
Comment (slontis) [inline providers/fips/self_test_kats.c:768]: I would prefer this to be a bit more flexible... i.e pass in the data pointers to this function. rather than hardwiring it..

File: providers/fips/self_test_kats.c
Comment (slontis) [inline providers/fips/self_test_kats.c:744]: Data should be in the .inc and be a structure that can be included for the test, in a data driven manner..

File: providers/fips/self_test_kats.c
Comment (paulidale) [reply providers/fips/self_test_kats.c:768]: It's called from exactly one place.  Why would it need any more flexibility at this stage?

File: providers/fips/self_test.c
Comment (paulidale) [reply providers/fips/self_test.c:183]: Wrong C file.  This one doesn't include the .inc file.  I think the HMAC needs to have a KAT before the integrity check is done, so here is the spot.

File: providers/fips/self_test_kats.c
Comment (slontis) [inline providers/fips/self_test_kats.c:793]: Why do you use HASH-DRBG?

File: crypto/dsa/dsa_key.c
Comment (slontis) [reply crypto/dsa/dsa_key.c:96]: You mean before it uses the buffer?

**Review by slontis — COMMENTED** [review]
Tests need to be added for these new shiny things.
See 03-testfipsinstall.t (corruption tests). Integrity check failure test also needed.
Failure during the rand change tests also should be done if not already called.

**Comment by slontis** [comment]
Documentation related to fipsinstall will need to be updated to reflect that we now need to run 'always'

**Review by slontis — COMMENTED** [review]
You may want to collapse some related commits into one. In particular the code that changes rand to public API's should be documented with why it was needed to expose them.

File: providers/fips/self_test_kats.c
Comment (slontis) [reply providers/fips/self_test_kats.c:768]: The rest of it was written that way.

File: test/recipes/03-test_fipsinstall.t
Comment (slontis) [reply test/recipes/03-test_fipsinstall.t:289]: Depends.. Is the code related to running once going to be removed from everywhere (inside the fipsinstall app also I mean).

File: providers/fips/self_test_kats.c
Comment (paulidale) [reply providers/fips/self_test_kats.c:793]: Ask KeyPair :)

It doesn't matter which is used.

File: test/recipes/03-test_fipsinstall.t
Comment (paulidale) [reply test/recipes/03-test_fipsinstall.t:289]: I'd think so.  That code is dead currently.

File: crypto/dsa/dsa_key.c
Comment (paulidale) [reply crypto/dsa/dsa_key.c:98]: Yep

File: crypto/ec/ec_key.c
Comment (paulidale) [reply crypto/ec/ec_key.c:279]: Worse, the tests fail with that change.

File: crypto/ec/ec_key.c
Comment (paulidale) [reply crypto/ec/ec_key.c:273]: It doesn't appear to need to be.

**Comment by paulidale** [comment]
> In particular the code that changes rand to public API's should be documented with why it was needed to expose them.

This is something I've wanted to add for a while, it's useful beyond FIPS.  So a comment about FIPS self tests needing it isn't required IMO.

**Comment by slontis** [comment]
> 

Perhaps, but it is not what motivated you to add it now right? it was required for the self tests. I just mean a comment in the commit not the general docs.

File: providers/fips/self_test_kats.c
Comment (beldmit) [reply providers/fips/self_test_kats.c:857]: We explicitly pass the random value and use it only in case we are in FIPS self-test mode

File: providers/fips/self_test.c
Comment (t8m) [reply providers/fips/self_test.c:179]: I do not think it matters.

Comment (slontis) [reply providers/fips/self_test.c:179]: It only matters if they were generated as dog food.. Not a known KAT.

File: test/recipes/03-test_fipsinstall.t
Comment (slontis) [reply test/recipes/03-test_fipsinstall.t:289]: I think that should be removed then at the same time (And the docs updated)

File: providers/fips/self_test.c
Comment (paulidale) [reply providers/fips/self_test.c:179]: ACVP produces myriad "unknown" KATs :grin:

**Comment by paulidale** [comment]
I think all comments are covered now.

I'll deal with the 2023 transitions et al in separate PRs.

File: crypto/ec/ec_key.c
Comment (slontis) [inline crypto/ec/ec_key.c:278]: Another idea would just be to do..
EC_PONT_set_to_infinity()... The main point of this is just to fail the test.. Dont really care about the buffer..

File: crypto/ec/ec_key.c
Comment (slontis) [reply crypto/ec/ec_key.c:547]: ??

File: doc/man1/openssl-fipsinstall.pod.in
Comment (slontis) [inline doc/man1/openssl-fipsinstall.pod.in:169]: since FIPS 140-3 removed this capability.

File: include/openssl/fips_names.h
Comment (slontis) [inline include/openssl/fips_names.h:33]: state why for these

File: providers/fips/self_test_data.inc
Comment (slontis) [inline providers/fips/self_test_data.inc:110]: thanks :)

File: providers/fips/self_test_kats.c
Comment (slontis) [inline providers/fips/self_test_kats.c:703]: doing these as OSSL_PARAM might be nicer..

Comment (paulidale) [reply providers/fips/self_test_kats.c:703]: That's what happened inside that function.  Why add additional mess here?

File: doc/man1/openssl-fipsinstall.pod.in
Comment (paulidale) [reply doc/man1/openssl-fipsinstall.pod.in:169]: It's coming back based on comments from @t-j-h, so no need.

File: providers/fips/self_test_kats.c
Comment (slontis) [reply providers/fips/self_test_kats.c:703]: I mean not using DRBG, but using new params setters that would be added to core_names.h and implemented in a pkey. (Up to you what method you use),

File: crypto/ec/ec_key.c
Comment (paulidale) [reply crypto/ec/ec_key.c:278]: This wouldn't really do the test.  I think I'll leave this as is: the self tests consistently flip a bit.  Otherwise, one could just return an error here and not call EC_POINT_cmp() at all.

File: crypto/ec/ec_key.c
Comment (paulidale) [reply crypto/ec/ec_key.c:547]: Dunno here.  Omiting the IG reference also addresses your point :smirk:

Comment (slontis) [reply crypto/ec/ec_key.c:547]: Please ask keypair - I couldnt find what this referred to..

File: providers/fips/self_test_kats.c
Comment (paulidale) [reply providers/fips/self_test_kats.c:703]: That ends up pretty ugly.  The RNGs are accessed rather deep down and intercepting things too high up isn't testing the key gen.

I'm thinking about setting up a separate libctx for these test cases, although I'm not sure how workable it is at this point.

I can verify that the DRBGs are restored correctly outside of the tests if extra assurance is required.

Comment (t8m) [reply providers/fips/self_test_kats.c:703]: > I can verify that the DRBGs are restored correctly outside of the tests if extra assurance is required.

Please do.

**Review by t8m — CHANGES_REQUESTED** [review]

File: providers/fips/self_test_kats.c
Comment (paulidale) [reply providers/fips/self_test_kats.c:703]: Done, FWIW.

File: crypto/ec/ec_key.c
Comment (paulidale) [reply crypto/ec/ec_key.c]: I think it's actually safe since BN_bin2bn() can't fail because the correct amount of memory is already there, but I've fixed it.

File: providers/fips/self_test_kats.c
Comment (paulidale) [reply providers/fips/self_test_kats.c:857]: I've added verification the the TEST-RAND has been replaced after the tests are run.  This should avoid this possibility.

File: providers/fips/self_test.c
Comment (t8m) [inline providers/fips/self_test.c:413]: Typo: restored

Comment (paulidale) [reply providers/fips/self_test.c]: Fixed

**Review by t8m — APPROVED** [review]

File: crypto/ec/ec_key.c
Comment (slontis) [inline crypto/ec/ec_key.c:386]: extra blankline

File: crypto/ec/ec_key.c
Comment (slontis) [reply crypto/ec/ec_key.c:548]: Wont hold it up on this - but it would be good to know where this came from.

File: include/openssl/self_test.h
Comment (slontis) [inline include/openssl/self_test.h:33]: Is that comment right?

**Review by slontis — COMMENTED** [review]
Have the public API changes been added to master?
They probably need to be done there before this PR can be merged.

**Comment by paulidale** [comment]
I was planning on merging this to 3.1 and cherry-picking to master which is kind of backwards.

**Comment by slontis** [comment]
Prefer we dont start doing this :)

**Review by slontis — APPROVED** [review]
LGTM - (I hate it when you force push)

**Comment by slontis** [comment]
Subject to being merged to master.. @t8m should reapprove also.

File: doc/man1/openssl-fipsinstall.pod.in
Comment (t-j-h) [inline doc/man1/openssl-fipsinstall.pod.in:188]: No it doesn't - and it shouldn't. This is a users choice to make.

File: apps/fipsinstall.c
Comment (t-j-h) [inline apps/fipsinstall.c:563]: This statement does not belong here like this.
We are talking about "requirements" that actually say what is being stated here.
The "best" we can do is direct the user to the "Security Policy" where they need to confirm their use is within the scope of the Security Policy. 

No definitive statements like this should exist.

And is this an error or not - it says it is a warning - but doesn't the goto end make it an error (haven't checked the context). It should not be an error.

**Comment by t-j-h** [comment]
Just placing a temporary hold that I'll release once comments are addressed - this should not be merged as-is IMHO without those being addressed.

**Comment by paulidale** [comment]
For the benefit of @slontis #19510 is live.

File: doc/man1/openssl-fipsinstall.pod.in
Comment (paulidale) [reply doc/man1/openssl-fipsinstall.pod.in:188]: The underlying code no longer support this mode of operation.  The self tests always run.

Comment (t-j-h) [reply doc/man1/openssl-fipsinstall.pod.in:188]: They should not always run. We should have the same options we have under the existing module. You can change the default behaviour if you wish - but the same handling should remain.

**Comment by paulidale** [comment]
Revert the change of behaviour.

**Review by t8m — APPROVED** [review]

File: util/libcrypto.num
Comment (t8m) [inline util/libcrypto.num:5432]: The version numbers are wrong here. They should be 3_1_0.

**Comment by paulidale** [comment]
@t-j-h, comments addressed.  Hold?

**Review by t-j-h — APPROVED** [review]

**Comment by t-j-h** [comment]
@paulidale thanks for reminder - looks good - approved and hold lifted

**Review by slontis — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by paulidale** [comment]
Merged, thanks.

**Pull request closed without merging**
